lang/src/accounts/interface.rs

```diff
@@ -0,0 +1,137 @@
+//! Type validating that the account is one of a set of given Programs
+
+use crate::accounts::program::Program;
+use crate::error::{Error, ErrorCode};
+use crate::{
+    AccountDeserialize, Accounts, AccountsExit, CheckId, Key, Result, ToAccountInfos,
+    ToAccountMetas,
+};
+use solana_program::account_info::AccountInfo;
+use solana_program::instruction::AccountMeta;
+use solana_program::pubkey::Pubkey;
+use std::collections::{BTreeMap, BTreeSet};
+use std::ops::Deref;
+
+/// Type validating that the account is one of a set of given Programs
+///
+/// The `Interface` wraps over the [`Program`](crate::Program), allowing for
+/// multiple possible program ids. Useful for any program that implements an
+/// instruction interface. For example, spl-token and spl-token-2022 both implement
+/// the spl-token interface.
+///
+/// # Table of Contents
+/// - [Basic Functionality](#basic-functionality)
+/// - [Out of the Box Types](#out-of-the-box-types)
+///
+/// # Basic Functionality
+///
+/// Checks:
+///
+/// - `expected_programs.contains(account_info.key)`
+/// - `account_info.executable == true`
+///
+/// # Example
+/// ```ignore
+/// #[program]
+/// mod my_program {
+///     fn set_admin_settings(...){...}
+/// }
+///
+/// #[account]
+/// #[derive(Default)]
+/// pub struct AdminSettings {
+///     ...
+/// }
+///
+/// #[derive(Accounts)]
+/// pub struct SetAdminSettings<'info> {
+///     #[account(mut, seeds = [b"admin"], bump)]
+///     pub admin_settings: Account<'info, AdminSettings>,
+///     #[account(constraint = program.programdata_address()? == Some(program_data.key()))]
+///     pub program: Interface<'info, MyProgram>,
+///     #[account(constraint = program_data.upgrade_authority_address == Some(authority.key()))]
```

Comment:
I'm not sure if the `Some()`s are necessary here? `authority` isn't an optional account so `authority.key()` will return a pubkey.

Reply:
`programdata_address()` on the `Program` type exposes an `Option<Pubkey>` for the address that holds the actual program data.  It's an `Option` because it doesn't exist for non-upgradeable programs. Since `Interface` can be dereferenced as a `Program`, it implicitly inherits the `programdata_address()` function.

And the `upgrade_authority_address` on the `ProgramData` type is an `Option<Pubkey>`, since an upgradeable program may have relinquished upgrade authority, making it `None`.

Reply:
Ah I get it. Cool, I've never really tried to deserialize those accounts. For some reason I thought you might be using the `Some()` because of the new `Option<Account<T>>`  optional account types that were introduced. Which coincidentally we also might need to make sure that `Option<Interface>` and `Option<InterfaceAccount>` still work.

Reply:
> Which coincidentally we also might need to make sure that Option<Interface> and Option<InterfaceAccount> still work.

Ah good call, I can add a test for that. It should work since they implement `Accounts`, which is what `Option` is bound on.

> `program.programdata_address()?` the question mark would unwrap the Option<Pubkey> though right, or is just a second layer of wrapper?

`programdata_address()` returns `Result<Option<Pubkey>>`, so the question mark just resolves the first layer, while still keeping it an `Option<Pubkey>`

Reply:
Added a test in [7d1ee1b](https://github.com/coral-xyz/anchor/pull/2386/commits/7d1ee1bf4f7aae9efc9e3d9b56b21daa7335e50f)